container/Dockerfile

```diff
@@ -1,18 +1,76 @@
-RUN mkdir /app
-ADD . /app
-WORKDIR /app
-RUN make
-ENTRYPOINT ["/bin/bash", "/opt/start.sh"]
+# Builder image contains header files and additional dependencies necessary to
+# generate wheel files.
+FROM debian:stable-slim AS builder
+
+LABEL org.opencontainers.image.authors=hannu.valtonen@aiven.io \
+      org.opencontainers.image.url=https://karapace.io \
+      org.opencontainers.image.documentation=https://github.com/aiven/karapace/ \
+      org.opencontainers.image.source=https://github.com/aiven/karapace/ \
+      org.opencontainers.image.vendor=Aiven \
+      org.opencontainers.image.licenses=Apache-2.0
+
+# Build dependencies that need to be installed:
+# - git: Used to install dependencies directly from their public repos (release
+#   not on PyPI).
+# - python3-devel: Python .h files, used to compile C extensions (e.g. multidict)
+#
+# Build dependencies that need to be installed because of `--no-install-recommends`:
+# - gcc: g++ and gcc to compile C extensions
+# - python3-wheel: Library to generate .whl files
+# - python3-setuptools: Packaging library
+#
+RUN apt-get update && \
+    apt-get -y install --no-install-recommends git python3-dev python3-pip python3-setuptools python3-wheel gcc && \
+    rm -rf /var/lib/apt/lists/*
+
+# Copy the requirements.txt and generate wheels for each dependency. Using a
+# separate command to use layer caching.
+#
+# Note: the requirements.txt is pinned, if any of the dependencies is updated
+# the cache will be invalidated and the image regenerated, which is the
+# intended behavior.
+#
+COPY ./requirements.txt /build/
+RUN pip3 wheel --requirement /build/requirements.txt --wheel-dir /build/dependencies-wheels
+
+COPY . /build/karapace-repo
+RUN pip3 wheel --no-deps /build/karapace-repo --wheel-dir /build/karapace-wheel
+
+# Karapace image.
+FROM debian:stable-slim AS karapace
+
+# Labels must be redefined beucase the base image is debian
+LABEL org.opencontainers.image.authors=hannu.valtonen@aiven.io \
+      org.opencontainers.image.url=https://karapace.io \
+      org.opencontainers.image.documentation=https://github.com/aiven/karapace/ \
+      org.opencontainers.image.source=https://github.com/aiven/karapace/ \
+      org.opencontainers.image.vendor=Aiven \
+      org.opencontainers.image.licenses=Apache-2.0
+
+RUN groupadd --system karapace && \
+    useradd --system --gid karapace karapace && \
+    mkdir /opt/karapace /var/log/karapace && \
+    chown --recursive karapace:karapace /opt/karapace /var/log/karapace
+
+RUN apt-get update && \
+    apt-get -y install --no-install-recommends python3-pip && \
+    rm -rf /var/lib/apt/lists/*
+
+COPY --from=builder /build/dependencies-wheels/*.whl /build/dependencies-wheels/
+RUN pip3 install --no-deps /build/dependencies-wheels/*.whl && rm -rf /build/dependencies-wheels/
+
+COPY --from=builder /build/karapace-wheel/*.whl /build/karapace-wheel/
+RUN pip3 install --no-deps /build/karapace-wheel/*.whl && rm -rf /build/karapace-wheel/
```

Comment:
It is also option to have three stages, where one is builder, second is for installing & generating runtime dependencies, and then final image just copies from runtime-deps images without installing pip or anything into resulting container. That would make the resulting image smaller (not containing pip).

Reply:
> then final image just copies from runtime-deps images

it is worth it though? what about I just delete the pip package before the commit?

Reply:
Adding:

```Dockerfile
RUN apt-get -y remove --prune python3-pip
```
The new image had size `53572016`, against `53583060`. Doesn't seem worth it

Reply:
A later command cannot remove files from underlying layer, so the way you tested it basically has no effect. The underlying layer still contains pip and pulling docker image would download that layer as part of the container image.

Regardless of that, I'm fine having pip included in the image, but if you want to remove pip from the image, it cannot be done that way adding that removal command later.

Reply:
Ah, you're right, I forgot that the format included all layers.